### PR TITLE
lint files by git diff

### LINKED ISSUE #150: May be bug that forget to write `()`
[test/commit.js#152](https://github.com/stevelacy/gulp-git/blob/master/test/commit.js#L152)
```js
gotData.should.be.false;
```

[test/rm.js#18](https://github.com/stevelacy/gulp-git/blob/master/test/rm.js#L18)
```js
exists.should.be.false;
```

[test/rm.js#37](https://github.com/stevelacy/gulp-git/blob/master/test/rm.js#L37)
```js
exists.should.be.false;
```

[test/status.js#21](https://github.com/stevelacy/gulp-git/blob/master/test/status.js#L21)
```js
exists.should.be.true;
```

### PR DESCRIPTION
**Pull request by gucong3000**
- Add new API `git.diff()`
- Add new API `git.catFile()`
- Use eslint instead of jshint
- Lint files that different from last release in `travis-ci`
- Add git-hook for lint file contents in git index
- Fixed #150

### CODE REVIEW

**Comment by gucong3000** [comment]
#147 #97

**Comment by gucong3000** [comment]
Here is a use case for new api:
https://github.com/gucong3000/gulp-git/blob/git_diff_flow/gulpfile.js

File: .travis.yml [hunk 1]
```diff
@@ -1,8 +1,14 @@
...
-  - "0.12"
+  - "stable"
+  - "6"
+  - "4"
 before_install:
-  -  git config --global user.email "you@example.com"
-  -  git config --global user.name "Your Name"
+  - git config --global user.email "you@example.com"
+  - git config --global user.name "Your Name"
+install_script:
```
Comment (stevelacy) [inline .travis.yml]: Please remove the `install_script` and `script` block. Travis should not include custom directives.

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+'use strict';
```
Comment (stevelacy) [inline gulpfile.js]: Please remove this file, it is not needed in the package.

File: lib/catFile.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+var through = require('through2');
+var gutil = require('gulp-util');
+var exec = require('child_process').exec;
+var escape = require('any-shell-escape');
+var stripBom = require('strip-bom-stream');
+var BufferStreams = require('bufferstreams');
+
+module.exports = function (opt) {
+  if (!opt) opt = {};
+  if (null == opt.stripBOM) opt.stripBOM = true;
```
Comment (stevelacy) [inline lib/catFile.js]: Please do `variable | comparator` as in `if (opt === null)`
Also use triple equals `===`, `!==` as the other files have.

File: lib/catFile.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  if (null == opt.buffer) opt.buffer = true;
+
+  var write = function(file, enc, cb) {
+    var hash = file.git && file.git.hash;
+
+    if(!hash) {
+      cb(null, file);
```
Comment (stevelacy) [inline lib/catFile.js]: `return cb(null, file);`

File: lib/catFile.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      return;
+    }
+
+    var cmd = 'git cat-file blob ' + hash;
+    var contents = exec(cmd, {cwd: file.cwd, maxBuffer: Number.MAX_VALUE}, function(err, stdout, stderr){
+      if (err) return cb(err);
+      if (contents) {
+        file.contents = contents;
+      }
+      cb(null, file);
```
Comment (stevelacy) [inline lib/catFile.js]: This `cb(null, file)` does not have a return, causing the following callbacks to get triggered

File: lib/catFile.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+var through = require('through2');
+var gutil = require('gulp-util');
+var exec = require('child_process').exec;
+var escape = require('any-shell-escape');
+var stripBom = require('strip-bom-stream');
+var BufferStreams = require('bufferstreams');
```
Comment (stevelacy) [inline lib/catFile.js]: Why bufferstreams instead of through?

File: package.json [hunk 7]
```diff
@@ -11,18 +11,25 @@
...
+    "through2": "^2.0.3",
+    "vinyl": "^2.0.1"
   },
   "devDependencies": {
-    "jshint": "^2.8.0",
-    "mocha": "^2.3.0",
+    "eslint": "^3.15.0",
+    "gulp": "^3.9.1",
+    "gulp-eslint": "^3.0.1",
+    "husky": "^0.13.1",
```
Comment (stevelacy) [inline package.json]: what is husky?
Also remove gulp-eslint after removing gulp and the gulpfile

File: package.json [hunk 8]
```diff
@@ -11,18 +11,25 @@
...
+    "mocha": "^3.2.0",
     "rimraf": "^2.4.3",
-    "should": "^7.1.0"
+    "should": "^11.2.0"
   },
   "scripts": {
-    "test": "mocha --reporter spec --timeout 6000 test/main.js && jshint ./index.js ./examples/. ./lib/. ./test/."
+    "test": "mocha --reporter spec --timeout 6000 test/main.js && gulp precommit",
+    "precommit": "gulp precommit"
```
Comment (stevelacy) [inline package.json]: Switch `gulp precommit` to `npm run lint` and add a new script like `test` called lint which will run eslint

File: test/catFile.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+'use strict';
+
+var fs = require('fs');
+var path = require('path');
+var rimraf = require('rimraf');
+var should = require('should');
+var gutil = require('gulp-util');
+var it = require('mocha').it;
```
Comment (stevelacy) [inline test/catFile.js]: `it` is inferred by scope through mocha, refer to the other tests. It should not be required

File: test/diff.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        dstMode: '100644',
+        srcHash: 'f128959',
+        dstHash: '1ade7a9',
+        status: 'M',
+        srcPath: 'package.json',
+        dstPath: 'package.json' } ]);
+
+      done();
+    }).on('error', done);
+
```
Comment (stevelacy) [inline test/diff.js]: whitespace

File: lib/catFile.js (hunk 6 above)
Comment (gucong3000) [inline lib/catFile.js]: It is not instead through.
It is use to read stream of Buffer.
```js
var contents = exec(cmd ...).stdout;
...
contents.pipe(new BufferStreams(function (err, buf, cb) {
  contents = buf;
  return cb(err, buf);
}));
```

File: lib/catFile.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    var cmd = 'git cat-file blob ' + hash;
+    var contents = exec(cmd, {cwd: file.cwd, maxBuffer: Number.MAX_VALUE}, function(err, stdout, stderr){
+      if (err) return cb(err);
+      if (contents) {
+        file.contents = contents;
+      }
+      cb(null, file);
+    }).stdout;
+
+    if (opt.stripBOM) {
```
Comment (stevelacy) [inline lib/catFile.js]: Striping the buffer order mark does seem like a way to narrow use case.

File: lib/catFile.js (hunk 6 above)
Comment (stevelacy) [inline lib/catFile.js]: There are now three stream libraries, through2, bufferstream and strip-bom-stream.
through2 can do what bufferstream does by just wrapping the `file.contents` in a `new Buffer()`

**Comment by stevelacy** [comment]
Excellent, thanks for making those changes

**Comment by stevelacy** [comment]
Could you send a PR with some docs for using diff and catfile?
Thanks

**Comment by bytestream** [comment]
git-gulp version 0.15 (this PR) seems to silently break compatibility with nodejs 0.10:
````
node_modules/gulp-git/node_modules/strip-bom-stream/index.js:2
const firstChunk = require('first-chunk-stream');
^^^^^
SyntaxError: Use of const in strict mode.
````

Upgrading to nodejs 4.x fixes the problem.

**Comment by stevelacy** [comment]
@bytestream published patch 1.15.1 which reverts the commits - going to publish a major change with the es6 dependencies

**Pull request merged**